### PR TITLE
Feature: Draft coin comparison charts

### PR DESCRIPTION
**Pull request by agreen254**
- display single dataset charts for bitcoin market cap and volume
- max number of data points per chart is capped at a threshold
- data compression is done via the LTTB algorithm (https://github.com/sveinn-steinarsson/flot-downsample/blob/master/jquery.flot.downsample.js)
- better styling of axes and legend to come

https://github.com/agreen254/crypto-app/assets/101513222/c232ca1a-e8c5-4125-9eac-356133e95e3c

### CODE REVIEW

File: app/api/v1/comparison/route.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  const bodyValidation = comparisonChartRequestSchema.safeParse(body);
+
+  if (!bodyValidation.success) {
+    throw new Error("Failed to validate the request.");
+  }
+
+  const { id, currency, days } = body;
+
+  const fetchHead = "https://api.coingecko.com/api/v3/coins/";
+  const fetchBody = `${id}/market_chart?vs_currency=${currency}&days=${days}`;
```
Comment (yoloOnTheBattlefield) [inline app/api/v1/comparison/route.ts]: const fetchHead = "https://api.coingecko.com/api/v3/coins";
  const fetchBody = `/${id}/market_chart?vs_currency=${currency}&days=${days}`;

we usually put the / in the next string

File: app/api/v1/comparison/route.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  const fetchTail = `&x_cg_demo_api_key=${process.env.COINGECKO_API_KEY}`;
+  const fetchURL = [fetchHead, fetchBody, fetchTail].join("");
+
+  let error: string | undefined;
+  const comparisonRes = await fetch(fetchURL, {
+    method: "GET",
+  }).catch((e) => (error = e));
```
Comment (yoloOnTheBattlefield) [inline app/api/v1/comparison/route.ts]: the default is get

Comment (yoloOnTheBattlefield) [reply app/api/v1/comparison/route.ts]: maybe don;t use `.catch` and make a `try catch` around your code

File: components/ComparisonCharts/ComparisonChartsWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+          res.error && (
+            <p
+              key={queryRequest.ids[idx]}
+              className="text-destructive mt-1 ml-1"
+            >
+              An unexpected error occurred for fetch {queryRequest.ids[idx]}:{" "}
+              {res.error.message}
+            </p>
+          )
+      )}
```
Comment (yoloOnTheBattlefield) [inline components/ComparisonCharts/ComparisonChartsWrapper.tsx]: use filter then map, and make sure you use a key in your map

**Review by yoloOnTheBattlefield — CHANGES_REQUESTED** [review]

**Comment by agreen254** [comment]
https://github.com/agreen254/crypto-app/assets/101513222/821a7aae-f48b-4017-ab08-00591edbf89a

**Pull request merged**